### DISCUSSION TITLE
Email verification message not being sent straight away

### DISCUSSION THREAD

**Discussion by LemonPie22** [filamentphp/filament#11569] [Help]
### Package

Panel builder

### Package Version

v3.2.34

### How can we help you?

When I enable email verification the user sees the correct prompt when they log in:

<img width="416" alt="Screenshot 2024-02-25 at 18 09 46" src="https://github.com/filamentphp/filament/assets/107488019/7b3e2e18-5eaa-4e60-ae37-0632ca457f58">

But the email is never sent through first time so the user always has to click the resend link.

As soon as they do this the whole process works perfectly.

My `AdminPanelProvider` class:
```
<?php

namespace App\Providers\Filament;

use App\Filament\Pages\EditProfile;
use Filament\Http\Middleware\Authenticate;
use Filament\Http\Middleware\DisableBladeIconComponents;
use Filament\Http\Middleware\DispatchServingFilamentEvent;
use Filament\Navigation\NavigationGroup;
use Filament\Pages;
use Filament\Panel;
use Filament\PanelProvider;
use Filament\Support\Colors\Color;
use Filament\Widgets;
use Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse;
use Illuminate\Cookie\Middleware\EncryptCookies;
use Illuminate\Foundation\Http\Middleware\VerifyCsrfToken;
use Illuminate\Routing\Middleware\SubstituteBindings;
use Illuminate\Session\Middleware\AuthenticateSession;
use Illuminate\Session\Middleware\StartSession;
use Illuminate\View\Middleware\ShareErrorsFromSession;
use Leandrocfe\FilamentApexCharts\FilamentApexChartsPlugin;

class AdminPanelProvider extends PanelProvider
{
    public function panel(Panel $panel): Panel
    {
        return $panel
            ->brandLogo(asset('images/sp_logo.webp'))
            ->darkModeBrandLogo(asset('images/sp_logo_dark.png'))
            ->brandLogoHeight('3.5rem')
            ->default()
            ->id('admin')
            ->path('admin')
            ->login()
            ->passwordReset()
            ->emailVerification()
            ->profile(EditProfile::class)
            ->databaseNotifications()
            ->plugins([
                FilamentApexChartsPlugin::make()
            ])
            ->colors([
                'primary' => Color::Amber,
            ])
            ->discoverResources(in: app_path('Filament/Resources'), for: 'App\\Filament\\Resources')
            ->discoverPages(in: app_path('Filament/Pages'), for: 'App\\Filament\\Pages')
            ->pages([
                Pages\Dashboard::class,
            ])
            ->discoverWidgets(in: app_path('Filament/Widgets'), for: 'App\\Filament\\Widgets')
            ->widgets([
//                Widgets\AccountWidget::class,
//                Widgets\FilamentInfoWidget::class,
            ])
            ->middleware([
                EncryptCookies::class,
                AddQueuedCookiesToResponse::class,
                StartSession::class,
                AuthenticateSession::class,
                ShareErrorsFromSession::class,
                VerifyCsrfToken::class,
                SubstituteBindings::class,
                DisableBladeIconComponents::class,
                DispatchServingFilamentEvent::class,
            ])
            ->authMiddleware([
                Authenticate::class,
            ]);
    }
}
```

Any ideas?

**Comment by leandrocfe** [comment]
Did you implement [MustVerifyEmail](https://laravel.com/docs/10.x/verification#model-preparation) interface?

**Reply by LemonPie22** [reply]
Yes I did. This is my User model:
```
<?php

namespace App\Models;

use Filament\Models\Contracts\FilamentUser;
use Filament\Models\Contracts\HasAvatar;
use Filament\Panel;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Relations\HasOneThrough;
use Illuminate\Foundation\Auth\User as Authenticatable;
use Illuminate\Notifications\Notifiable;
use Laravel\Sanctum\HasApiTokens;
use Illuminate\Database\Eloquent\SoftDeletes;
use OwenIt\Auditing\Contracts\Auditable;
use Spatie\Permission\Traits\HasRoles;
use Illuminate\Contracts\Auth\MustVerifyEmail;

class User extends Authenticatable implements FilamentUser, Auditable, MustVerifyEmail, HasAvatar
{
    use HasApiTokens;
    use HasFactory;
    use SoftDeletes;
    use Notifiable;
    use \OwenIt\Auditing\Auditable;
    use HasRoles;

    public function getFilamentAvatarUrl(): ?string
    {
        // Display user image, if set
        if (empty($this->image)) {
            return null;
        }
        return asset('storage/' . $this->image);
    }

    /**
     * The attributes that are mass assignable.
     *
     * @var array<int, string>
     */
    // protected $fillable = [
    //     'name',
    //     'email',
    //     'password',
    // ];
    protected $guarded = [];

    /**
     * The attributes that should be hidden for serialization.
     *
     * @var array<int, string>
     */
    protected $hidden = [
        'password',
        'remember_token',
    ];

    /**
     * The attributes that should be cast.
     *
     * @var array<string, string>
     */
    protected $casts = [
        'email_verified_at' => 'datetime',
    ];
```

**Reply by adriaardila** [reply]
Having this issue too. Did you manage to fix it? @LemonPie22

**Reply by LemonPie22** [reply]
No. I'm going to try changing my QUEUE_CONNECTION as advised by @ranaabdulrauf

**Reply by adriaardila** [reply]
I have redis in my local environment as QUEUE_CONNECTION, for production environment I use Amazon SQS and I still have the same problem.

**Reply by authanram** [reply]
Same issue here. I tried the QUEUE_CONNECTION advice from @ranaabdulrauf without success.

**Reply by aqilmujahidiin** [reply]
me too having this issue, user must click resend it to send email

**Reply by evdmeer** [reply]
I am running into the same issue as well.
All mails are being sent as supposed, but the email verification mail will only be sent if the user clicks the resend link.

**Comment by ranaabdulrauf** [comment]
check you queue driver, it may be set to database or something else
QUEUE_CONNECTION="database" // redis

**Reply by LemonPie22** [reply]
I've tried `database`, `sync` and `redis`. It makes no difference.

**Comment by adriaardila** [comment]
The reason why verification emails are not being sent at first is because the verification mail sending is triggered on registration, not when you go into the email verification page. As you can see in vendor/filament/filament/src/Pages/Auth/Register.php, there is a function called "sendEmailVerificationNotification" that gets called right at the register method:

```
public function register(): ?RegistrationResponse
    {
        try {
            $this->rateLimit(2);
        } catch (TooManyRequestsException $exception) {
            $this->getRateLimitedNotification($exception)?->send();

            return null;
        }

        $user = $this->wrapInDatabaseTransaction(function () {
            $this->callHook('beforeValidate');

            $data = $this->form->getState();

            $this->callHook('afterValidate');

            $data = $this->mutateFormDataBeforeRegister($data);

            $this->callHook('beforeRegister');

            $user = $this->handleRegistration($data);

            $this->form->model($user)->saveRelationships();

            $this->callHook('afterRegister');

            return $user;
        });

        event(new Registered($user));

        $this->sendEmailVerificationNotification($user);

        Filament::auth()->login($user);

        session()->regenerate();

        return app(RegistrationResponse::class);
    }
```

You can implement your own logic and trigger the send by simply doing this using the user model:

$user->sendEmailVerificationNotification();

So basically, it works correctly for new registrations after you have implemented MustVerifyEmail, but for older users that were created without the trait they will have this problem, so maybe you can try a workaround for old users that triggers the method manually.

**Reply by evdmeer** [reply]
Aha, that explains my specific issue.
We import users from external systems and then send them an invitation mail by manually triggering the reset password flow with a custom email.
When they click that link. they get to choose their password (as we can't / don't want to migrate that from the external systems).

After logging in with their new password they got redirected to the verify email page, which did not send an email.
It makes sense that they never received it, since we bypassed the normal registration flow by manually adding the users.

It's good to know this is by design and we have just disabled the email verification for externally imported users (they receive their invitation mail via email anyways, so the email verification was superfluous anyways).
